### PR TITLE
[WIP] Write tests for issue API

### PR DESCRIPTION
**Pull request by andreynering**
Please don't merge yet. Opened for discussion on testing approach.

For context see #63 

Edit: to run these tests:

```
go test -v -tags sqlite ./routers/api/v1/repo/
```

### CODE REVIEW

File: modules/setting/setting.go [hunk 1]
```diff
@@ -481,8 +491,8 @@ func NewContext() {
 	}[Cfg.Section("time").Key("FORMAT").MustString("RFC1123")]
 
 	RunUser = Cfg.Section("").Key("RUN_USER").String()
-	// Does not check run user when the install lock is off.
-	if InstallLock {
+	// Does not check run user when the install lock is off or is running tests
+	if InstallLock && AppRunMode != RUN_MODE_TEST {
```
Comment (makhov) [inline modules/setting/setting.go:512]: Change code for testing is a bad practice

File: models/models.go [hunk 2]
```diff
@@ -57,6 +57,10 @@ var (
 	EnableTiDB    bool
 )
 
+func Database() *sql.DB {
```
Comment (makhov) [inline models/models.go:60]: And now we can execute SQL queries from handlers?

File: routers/api/v1/repo/issue_test.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+func TestIssuesAPI(t *testing.T) {
+	Convey("Given the issue API", t, func() {
+		testutil.TestGlobalInit()
+
+		Convey("It should return issues from the index", func() {
+			testutil.PrepareTestDatabase()
+
+			w, r := testutil.NewTestContext("GET", "/api/v1/repos/user1/foo/issues", "", nil, "1")
+			testutil.ServeHTTP(w, r)
+			So(w.Code, ShouldEqual, http.StatusOK)
```
Comment (makhov) [inline routers/api/v1/repo/issue_test.go]: Test which test nothing

**Review by makhov — CHANGES_REQUESTED** [review]

File: modules/setting/setting.go (hunk 1 above)
Comment (andreynering) [reply modules/setting/setting.go:512]: I agree. This is a WIP, we should abstract things using interfaces. I am open for suggestions on how to do it right.

**Comment by tboerger** [comment]
@makhov glad that you are back :)

File: routers/api/v1/repo/issue_test.go (hunk 3 above)
Comment (andreynering) [reply routers/api/v1/repo/issue_test.go]: Well, it tests if `/api/v1/repos/user1/foo/issues` returns status 200.

Comment (makhov) [reply routers/api/v1/repo/issue_test.go]: I see, but it useless info. `/api/v1/repos/user1/foo/issues` may returns 200 and `{"foo": "bar"}` in response body. Or `{"oops": "something went wrong"}`. Or `Fuck the json`, etc…

File: modules/setting/setting.go (hunk 1 above)
Comment (makhov) [reply modules/setting/setting.go:512]: Ok, thank you. It's very important to write testable code 👍

File: modules/setting/setting.go [hunk 4]
```diff
@@ -263,6 +269,10 @@ var (
 	HasRobotsTxt bool
 )
 
+func GiteaPath() string {
```
Comment (bkcsoft) [inline modules/setting/setting.go:272]: Exported functions should have be commented :smile:

File: testdata/app_test.ini [hunk 5]
```diff
@@ -0,0 +1,31 @@
+# THIS FILE EXISTS TO BE USED TO UNIT TESTS
+# DON'T CHANGE IT. THIS IS NOT THE FILE GOGS WILL USE IN PRODUCTION
```
Comment (bkcsoft) [inline testdata/app_test.ini:2]: `GOGS` ? :trollface:

File: testdata/app_test.ini [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+APP_NAME = Gogs: Go Git Service
```
Comment (bkcsoft) [inline testdata/app_test.ini:4]: Same as above

**Review by bkcsoft — CHANGES_REQUESTED** [review]

**Comment by bkcsoft** [comment]
As I'm a big promoter of unit-tests, I'd prefer if we'd start with the model-tests instead. That way we get rid of a bunch of edge-cases before going for e2e-tests (integration, like these ones). It's also extremely hard to force edge-cases from this level, and the amount of tests would blow out of proportions. Testing models is easier since we know what goes in an out, and we don't have to force edge-cases. I also suggest that we use in-memory sqlite database

**Comment by andreynering** [comment]
So this need more discussion...

**Pull request closed without merging**